torchvision/models/unet.py

```diff
@@ -0,0 +1,154 @@
+import torch
+import torch.nn as nn
+
+
+__all__ = ['UNet', 'unet8', 'unet13', 'unet18', 'unet23', 'unet28', 'unet33']
+
+
+def double_conv(in_channels, out_channels):
+    return nn.Sequential(
+        nn.Conv2d(in_channels, out_channels, 3),
+        nn.ReLU(inplace=True),
+        nn.Conv2d(out_channels, out_channels, 3),
+        nn.ReLU(inplace=True)
+    )
+
+
+def center_crop(img, output_size):
+    _, _, h, w = img.size()
+    _, _, th, tw = output_size
+    i = (h - th) // 2
+    j = (w - tw) // 2
+    return img[:, :, i:i + th, j:j + tw]
+
+
+class Contract(nn.Module):
+
+    def __init__(self, in_channels, out_channels, dropout=False, p=0.5):
+        super(Contract, self).__init__()
+        assert in_channels < out_channels
+
+        self.pool = nn.MaxPool2d(2)
+        self.conv = double_conv(in_channels, out_channels)
+        self.drop = None
+
+        if dropout:
+            self.drop = nn.Dropout2d(p=p)
+
+    def forward(self, x):
+        x = self.pool(x)
+        x = self.conv(x)
+
+        if self.drop is not None:
+            x = self.drop(x)
+
+        return x
+
+
+class Expand(nn.Module):
+
+    def __init__(self, in_channels, out_channels):
+        super(Expand, self).__init__()
+        assert in_channels > out_channels
+
+        self.upconv = nn.ConvTranspose2d(in_channels, out_channels, 2, 2)
+        self.relu = nn.ReLU(inplace=True)
+        self.conv = double_conv(in_channels, out_channels)
+
+    def forward(self, x, out):
+        x = self.upconv(x)
+        x = self.relu(x)
+
+        out = center_crop(out, x.size())
+        x = torch.cat([out, x], 1)
+
+        x = self.conv(x)
+
+        return x
+
+
+class UNet(nn.Module):
+    """`U-Net <https://arxiv.org/pdf/1505.04597.pdf>`_ architecture.
+
+    Args:
+        in_channels (int, optional): number of channels in input image
```

Comment:
So the reason I chose `in_channels=1` as the default for U-Net is because this is how the original U-Net paper is modeled, using a single channel grayscale microscope imagery dataset (see #900). The application I needed it for was actually 4-channel microscope imagery, but unfortunately PIL doesn't support this (see #882). If we decide to pretrain this on COCO/Pascal I'm fine with switching the default to 3-channel.